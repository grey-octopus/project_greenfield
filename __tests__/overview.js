import React from "react";
import { mount, shallow, render } from "enzyme";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import updateAverageRating from "../src/overview/actions/updateAverageRating.js";
import averageRatingReducer from "../src/overview/reducers/averageRatingReducer.js";
import fetchMock from "fetch-mock";
import StarRatingContainer from "../src/overview/containers/StarRatingContainer.jsx";
import ProdOverviewContainer from "../src/overview/containers/ProdOverviewContainer.jsx";
import ProdOverview from '../src/overview/components/ProdOverview.jsx'
import fetchProductInfo from "../src/overview/actions/fetchProductInfo.js";
import fetchProductInfoReducer from "../src/overview/reducers/fetchProductInfoReducer.js";
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history'
import { Provider } from 'react-redux'
import toJson from 'enzyme-to-json'
import fetchProductOverview from '../src/overview/actions/fetchProductOverview.js'

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe("actions", () => {
  describe('updateAverageRating', () => {
    it ('creates action with average rating', () => {
      const store = mockStore({ prodId: 1 })
      const expectedActions = [{ type: 'UPDATE_AVERAGE_RATING', numOfRatings: 7, payload: '4.00' }]
      return store.dispatch(updateAverageRating(1)).then(() => {
        expect(store.getActions()).toEqual(expectedActions)
      })
    })
  
    it ('should round payload to nearest quarter', () => {
      const store = mockStore({ prodId: 2})
      const expectedActions = [{ type: 'UPDATE_AVERAGE_RATING', numOfRatings: 5, payload: '3.50'}]
      return store.dispatch(updateAverageRating(2)).then(() => {
        expect(store.getActions()).toEqual(expectedActions)
      })
    })
  })

  it('fetchProductInfo', () => {
    const store = mockStore({ prodId: 1 });
    const expectedActions = [
      {
        type: "FETCH_PROD_INFO",
        title: "Camo Onesie",
        category: "Jackets"
      }
    ];
    return store.dispatch(fetchProductInfo(1)).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('fetchProductOverview', () => {
    const store = mockStore();
    const expectedActions = [
      {
        type: 'FETCH_PROD_OVERVIEW',
        slogan: 'Blend in to your crowd',
        description: 'The So Fatigues will wake you up and fit you in. This high energy camo will have you blending in to even the wildest surroundings.',
        features: [
          {
              "feature": "Buttons",
              "value": "Brass"
          }
        ]
      }
    ]
    return store.dispatch(fetchProductOverview(1)).then(() => {
      expect(store.getActions()).toEqual(expectedActions)
    })
  })
});

describe('reducers', () => {
  describe('averageRatingReducer', () => {
    it ('should handle updateAverageRating', () => {
      expect(averageRatingReducer([], { type: 'UPDATE_AVERAGE_RATING', payload: '4.00', numOfRatings: 5 })).toEqual({ numOfRatings: 5, averageRating: '4.00' })
    })
  })
  
  describe("fetchProductInfoReducer", () => {
    it("FETCH_PROD_INFO", () => {
      expect(
        fetchProductInfoReducer([], {
          type: "FETCH_PROD_INFO",
          title: "Fancy Jeans",
          category: "Jeans"
        })
      ).toEqual({
        title: "Fancy Jeans",
        category: "Jeans"
      });
    });
  
    it('FETCH_PROD_OVERVIEW', () => {
      expect(fetchProductInfoReducer([], {
        type: 'FETCH_PROD_OVERVIEW',
        category: "Jeans",
        description: "These jeans are fancy",
        slogan: "no",
        features: []
      })).toEqual({
        description: "These jeans are fancy",
        slogan: "no",
        features: []
      })
    })
  });
})



describe("<ProdOverview />", () => {
  it("renders with required props", () => {
    const props = {
      slogan: 'That\'s fair',
      description: 'Just a cool guy',
      features: ['test']
    }
    const history = createMemoryHistory()
    const wrapper = shallow(<Router history={history} path='/:prodId'><ProdOverview {...props} /></Router>)
    expect(toJson(wrapper.find('ProdOverview'))).toMatchSnapshot()
    expect(wrapper.find('ProdOverview').prop('slogan')).toEqual('That\'s fair')
    expect(wrapper.find('ProdOverview').prop('description')).toEqual('Just a cool guy')
    expect(wrapper.find('ProdOverview').prop('features')).toBeTruthy()
  });
});

// describe('<StarRatingContainer />', () => {
//   it ('should set averageRating property', () => {
//     const store = mockStore({ prodId: 1 })
//     const root = shallow(<StarRatingContainer store={store} />)
//     return store.dispatch(updateAverageRating(1)).then(() => {
//       expect(root.find('StarRating').prop)
//     })
//   })
// })
