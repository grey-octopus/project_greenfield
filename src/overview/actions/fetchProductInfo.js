import axios from "axios"

const fetchProductInfo = prodId => {
  return dispatch => {
    return axios.get(`http://3.134.102.30/products/${prodId}`).then(data => {
      return dispatch({ 
        type: 'FETCH_PROD_INFO', 
        title: data.data.name,
        category: data.data.category, 
        description: data.data.description, 
        slogan: data.data.slogan,
        features: data.data.features
      })
    })
  }
}

export default fetchProductInfo