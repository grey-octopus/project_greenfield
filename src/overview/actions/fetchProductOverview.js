import axios from 'axios';

const fetchProductOvervew = prodId => {
  console.log('DISPATCHED')
  return dispatch => {
    return axios.get(`http://3.134.102.30/products/${prodId}`).then(data => {
      return dispatch({
        type: 'FETCH_PROD_OVERVIEW',
        slogan: data.data.slogan,
        description: data.data.description,
        features: data.data.features
      });
    });
  };
};

export default fetchProductOvervew;