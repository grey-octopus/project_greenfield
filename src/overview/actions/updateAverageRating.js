import axios from 'axios';
import { calculateAvgRating } from '../../helpers.js';

const updateAverageRating = prodId => {
  return dispatch => {
    return axios
      .get(`http://3.134.102.30/reviews/${prodId}/meta`)
      .then(meta => {
        const ratings = meta.data.ratings;
        try {
          return dispatch({
            type: 'UPDATE_AVERAGE_RATING',
            payload: calculateAvgRating(ratings)
          });
        } catch {
          return dispatch({
            type: 'UPDATE_AVERAGE_RATING',
            numOfRatings: 0,
            payload: 0
          });
        }
      });
  };
};

export default updateAverageRating;
