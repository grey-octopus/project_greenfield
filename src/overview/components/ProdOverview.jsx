import React, { useEffect } from 'react';
import { useParams } from 'react-router-dom'
import DescriptionContainer from '../containers/DescriptionContainer.jsx';
import FeaturesContainer from '../containers/FeaturesContainer.jsx'

const ProdOverview = props => {
  const { prodId } = useParams()
  useEffect(() => { props.fetchProductOverview(prodId) }, [prodId]);
  if (props.description && props.slogan && props.features) {
    return (
      <div id='prod-overview'>
        <DescriptionContainer />
        <FeaturesContainer />
      </div>
    )
  } else {
    return <div id="description-hidden"></div>;
  }
};

export default ProdOverview;
