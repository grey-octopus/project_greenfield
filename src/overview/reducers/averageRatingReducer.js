const averageRatingReducer = (state = [], action) => {
  switch (action.type) {
    case 'UPDATE_AVERAGE_RATING':
      return {...state, averageRating: action.payload}
    default:
      return state;
  }
}

export default averageRatingReducer