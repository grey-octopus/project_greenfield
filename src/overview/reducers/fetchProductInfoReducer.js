const fetchProductInfoReducer = (state = [], action) => {
  switch (action.type) {
    case 'FETCH_PROD_INFO':
      return {
        ...state,
        title: action.title,
        description: action.description,
        category: action.category,
        slogan: action.slogan,
        features: action.features
      };
    default:
      return state;
  }
};

export default fetchProductInfoReducer;
