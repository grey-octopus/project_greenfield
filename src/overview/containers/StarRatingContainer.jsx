import { connect } from 'react-redux'
import StarRating from '../components/StarRating.jsx'
import updateAverageRating from '../actions/updateAverageRating.js'

const mapStateToProps = state => {
  return {
    averageRating: state.averageRatingReducer.averageRating,
    prodId: state.fetchProductInfoReducer.prodId
  }
}

const mapDispatchToProps = dispatch => {
  return {
    updateAverageRating: prodId => dispatch(updateAverageRating(prodId))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(StarRating)