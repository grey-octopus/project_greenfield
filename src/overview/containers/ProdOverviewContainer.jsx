import ProdOverview from '../components/ProdOverview.jsx'
import { connect } from 'react-redux'

export default connect(state => ({ title: state.fetchProductInfoReducer.title, category: state.fetchProductInfoReducer.category }))(ProdOverview)