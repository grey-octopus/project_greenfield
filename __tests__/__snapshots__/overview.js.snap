// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<ContextProvider
  value={null}
>
  <Description
    description="testDescription"
    dispatch={[Function]}
    slogan="testSlogan"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;

exports[` 2`] = `
<ContextProvider
  value={null}
>
  <Features
    dispatch={[Function]}
    features="testFeature"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;

exports[` 3`] = `
<ContextProvider
  value={null}
>
  <Overview
    fetchProductInfo={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    title="testTitle"
  />
</ContextProvider>
`;

exports[` 4`] = `
<ContextProvider
  value={null}
>
  <ProdInfo
    category="testCategory"
    dispatch={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    title="testTitle"
  />
</ContextProvider>
`;

exports[` 5`] = `
<ContextProvider
  value={null}
>
  <ProdOverview
    description="testDescription"
    features="testFeature"
    fetchProductOverview={[Function]}
    slogan="testSlogan"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;

exports[` 6`] = `
<ContextProvider
  value={null}
>
  <StarRating
    averageRating="4.00"
    numOfRatings={3}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    updateAverageRating={[Function]}
  />
</ContextProvider>
`;

exports[`components <Description /> renders with required props 1`] = `
<div>
  <Description
    description="Just a cool guy"
    slogan="That's fair"
  />
</div>
`;

exports[`components <Features /> renders with required props 1`] = `
<div>
  <Features
    features={
      Array [
        Object {
          "feature": "feature",
          "value": "val",
        },
      ]
    }
  />
</div>
`;

exports[`components <ProdOverview /> renders with required props 1`] = `
<ProdOverview
  description="Just a cool guy"
  features={
    Array [
      "test",
    ]
  }
  slogan="That's fair"
/>
`;
